Amazon.QLDB.Driver/session/SessionPool.cs

```diff
@@ -141,9 +141,14 @@ internal QldbSession GetSession()
             }
         }
 
+        internal int AvailablePermit()
+        {
+            return this.poolPermits.CurrentCount;
+        }
+
         private void ReleaseSession(QldbSession session)
         {
-            if (session != null)
+            if (session != null && session.IsAlive())
```

Comment:
I wish we didn't have to have this logic here and instead encapsulate this in the error returned, but I think because of the way we've abstracted SessionPool and RetryHandler out it isn't easily feasible.

Reply:
I understand your concern. If we treat whether the invalid session as an error, it does make sense to have it the in the exception. However, in the current .Net implementation, the exceptions returning from QldbSession are mostly used for the retry logic, and the session pool is not interfering with that or handle exceptions. We can have the retry logic to feedback this session liveness information to the session pool, but I feel then it is less direct than the current implementation.

We actually have two kinds of 'invalid session' here. The first is the InvalidSessionException, which get popped up to the retry logic. We also have the invalid session where other exceptions would be popped up to retry logic, but since it seems to fail on calling abort, we would like not to return this to the session pool. They are not the same, and the second information is only useful to the session pool.

So the logic here is whether the session is alive becomes a state of the session. The session pool decide whether to return a session back to pool based on whether it is null and whether it is still alive. I think the beauty here is the session pool does not have to depend on the retry logic to make things right for it. If we decide to remove the retry logic, we can simply remove the lines 64 to 73 of SessionPool.cs, and replace it with `return session.Execute(func);`, the session pool would still be functional.